### PR TITLE
refactor: Port the `VmTraceDecoder` from Hardhat

### PR DESCRIPTION
**Pull request by Xanewok**
This ports of `VmTraceDecoder` from Hardhat; what's left now is the error inferrer and finally solidity stack trace logic :partying_face: 

What's more interesting, however, is the fix for the `MessageTrace` and others:
1. There's a `napi-rs` bug that incorrectly validates and converts an object that has a required property but can be explicitly undefined, e.g. `deployedContract: Bytecode | undefined`. That's the case for `CreateMessageTrace` and so this is worked around using an optional property, however...
2. With this, the `CallMessageTrace` is a strict superset of `CreateMessageTrace`. We need to be careful with the type order when using the N-API type `Either` to always attempt converting and validating the most specific type first, otherwise we may incorrectly convert into a less specific type, skipping some additional properties it may have.

Link to the companion commit: https://github.com/nomicFoundation/hardhat/commit/137cfb33a8868329ae4d3d3903f4c7bf8838a50f

### CODE REVIEW

File: crates/edr_napi/src/trace/message_trace.rs
Comment (Wodann) [inline crates/edr_napi/src/trace/message_trace.rs:48]: FYI, backwards compatibility for Hardhat's internals is not a requirement, so if it's easier, you can just adopt `Option<Uint8Array>` and change the Hardhat code.

File: crates/edr_napi/src/trace/message_trace.rs
Comment (Wodann) [inline crates/edr_napi/src/trace/message_trace.rs:69]: If this is now filled on the Rust side, is there a reason we can't define an N-API `Bytecode` type?

**Review by Wodann — COMMENTED** [review]
Overall LGTM. Some questions and comments

File: crates/edr_napi/src/trace/message_trace.rs
Comment (Xanewok) [reply crates/edr_napi/src/trace/message_trace.rs:69]: We do define and use `Bytecode` elsewhere already but there's some subtle nuance here:

1. We can't use `ClassInstance` here because it internally holds a `&'static mut T` so that's UB because we probably will often have traces that point to the same bytecode
2. I introduced the `ClassInstanceRef` that guards the access with a `RefCell`, however it's not possible to directly use it here as it'd require `ToNapiValue` and `FromNapiValue` implementation...
3. and while it might be feasible to implement it using `ToNapiValue` (it can pass the underlying object), however...
4. implementing `FromNapiValue` would be very tricky:
  a. the instances are shared, so on the Rust side are wrapped in `Rc` and internally uses `napi::Ref` for napi-side ref-counting of the underlying object, so...
  b. converting at the FFI boundary without a side-table would have to create new Rc'd objects; ignoring the potential perf drawbacks, it would also be incorrect because it would lead to multiple `RefCells` guarding the same memory access and...
  c. introducing the side-table means introducing a static map per JS class on the Rust side, hooking into the napi lifecycle to manage the objects on the Rust side once they're freed on the JS side, modifying the `ClassInstanceRef` helper to do the same on the Rust side...
  d. ...at which point we'd be reimplementing quite a bit of higher-level convenience code that arguably should be on the napi-rs side itself; hopefully, we won't need as much JS glue anymore once the port is complete.
  
Let me know if the context makes sense; I'm more than happy if we simplified this somehow but I didn't manage to find a way to make it both safe and convenient :cry:

**Review by fvictorio — APPROVED** [review]
LGTM, but let's wait for @Wodann's approval

File: crates/edr_napi/src/trace/message_trace.rs
Comment (Wodann) [reply crates/edr_napi/src/trace/message_trace.rs:69]: It sounds very complicated. I'll trust your judgement!

Comment (Xanewok) [reply crates/edr_napi/src/trace/message_trace.rs:69]: Thank you! I'll try to embed a bit more context and/or more explain these things more concisely next time :pray:

**Pull request merged**
